### DISCUSSION TITLE
Welcome to termdash Discussions!

### DISCUSSION THREAD

**Discussion by mum4k** [mum4k/termdash#279] [General]
## 👋 Welcome!
  We’re using Discussions as a place to connect with other members of our community. We hope that you:
  * Ask questions you’re wondering about.
  * Share ideas.
  * Engage with other community members.
  * Welcome others and are open-minded. Remember that this is a community we
  build together 💪.

**Comment by lefred** [comment]
Hello, 
I´ve created a simple container have a text widget.
What is the best method to update the text (clean and change it) every x seconds or capture a key pressed ?
Thank you.

**Reply by lefred** [reply]
Found it ;)

**Reply by mum4k** [reply]
Glad you were able to find what you needed @lefred, please don't hesitate to ask if you have any questions.

**Reply by lefred** [reply]
Thank you. 
Is there a way do hide or delete a widget in a container ?

**Reply by mum4k** [reply]
Interestingly we do have the `PlaceWidget` method on container, but don't have a corresponding `RemoveWidget`. We could either add it, or as a workaround you could use the [DynamicLayout](https://github.com/mum4k/termdash/wiki/Dynamic-layout) support to replace the entire container. You would achieve this by having a "root" container and a child container with the widget. To remove the widget you would update the root container, splitting it again and thus creating a new child container without the widget. Please let me know if this helps, or if you are interested in adding the `RemoveWidget` method.

**Reply by mum4k** [reply]
Sorry, I was wrong. We do have a `Clear` method, that should do what you are asking:

https://github.com/mum4k/termdash/blob/fc131ea4369c0e2d0d5a0a29dcd5d286cee5311c/container/options.go#L381-L391

**Reply by lefred** [reply]
Perfect, exactly what I was looking for, thank you !

**Reply by lefred** [reply]
Hello, 
Is there a way to update/change  the textinput.Label for an existing textinput widget ?
Thank you

**Reply by mum4k** [reply]
The `TextInput` widget doesn't currently support updating/changing any of its options. The only way to replace the label without adding new functionality is to create a new `TextInput` widget instance with the updated options and use it to replace the existing one in the container. If this doesn't work, it should be a relatively trivial code change to allow updating/changing of options on an existing `TextInput` widget. Does the former solution achieve what you need?

**Reply by lefred** [reply]
Hi, 
Thank you, yes the former solution worked as expected.
Having an option to update the input text would be nice to not have to rewrite the onSubmit... but I could replace that code into a different function for example...

**Reply by mum4k** [reply]
Agreed, it's only a workaround. If it becomes too much of a nuisance - feel free to open an issue and we can discuss how and who can implement a better solution.

**Reply by lefred** [reply]
If you are curious to see where I use your project: https://github.com/lefred/innotopgo

thank you for this work !

**Reply by mum4k** [reply]
That looks awesome! What a great way to learn Go. Please feel free to send a PR adding yourself to the list of projects that use Termdash in the main README if you would like to do so.

Out of curiosity - what did you use to create the tabular text representation inside the text widget? We may want to document that for other users, since Termdash still doesn't have the table widget.

**Reply by lefred** [reply]
Hi, 
Thx, I´m just using frmt.Sprintf in some functions.
Something that would be also very cool is to have the sparkline but horizontal, something like: 
```
LABEL1 : ##########
LABEL2: ####
LABEL3: #####################
```
cheers,

**Reply by lefred** [reply]
Do you know if it´s possible to flush the keyboard events queue ?
Thx !

**Reply by mum4k** [reply]
Would you be able to explain that a little bit more? It would be the best if you could help me to understand the problem the flush should be solving.

**Reply by lefred** [reply]
My problem is that I create a new with termdash.KeyboardSubscriber, then if a certain key is pressed, I "change" dashboard where I create a new one (new go function fetching new information every seconds, etc...) that new context also subscribe the keyboard events to be catch the key to exit completely the application or go back to the main dashboard.
That is working. However if I press keys, once I go back in the main dashboard, those keys are again processed. I would like to have the possibility when I close the second created to flush the keyboard queue.


The issue is explained also here: https://github.com/lefred/innotopgo/issues/16

**Reply by mum4k** [reply]
Thank you for providing the explanation. I suspect that flushing of the queue won't help here, because the delivery of the events is concurrent - some events might still be delivered between the container update and the flush.

This is an interesting problem and something we should solve. I have a couple of ideas, but first I need to understand the issue a bit better:

1) When you say some keys get delivered again - do they get delivered again to the `termdash.KeyboardSubscriber`, or does a copy of them get delivered to the updated container and some widgets in it? I suspect the latter, but please confirm.
2) Could you provide more details on what do the duplicate keys cause in your application? Ideally, can you point me to the code that defines the `termdash.KeyboardSubscriber` that processes the first copy of the key and the code that processes the duplicate keys that got delivered again?

This will help me understand what we are dealing with and I should be able to suggest some solutions afterwards.

**Reply by lefred** [reply]
In this part of the code I wait for the user to pres ESC (to exit) or backspace to close this dashboard: 

https://github.com/lefred/innotopgo/blob/9df1f34ccc073f4d27d9b4146a1a42f2a87d530d/innotop/memory.go#L244

If other keys are pressed I don´t want to do anything... however if somebody press a key like ´I´ , then when this dashboard is closed and we go back  in the main one:

https://github.com/lefred/innotopgo/blob/9df1f34ccc073f4d27d9b4146a1a42f2a87d530d/innotop/processlist.go#L466

the keys that were pressed previously (<i> in my example) will then be processed and the action linked to that key will be performed (display of yet another dashboard).

**Reply by mum4k** [reply]
Thank you for providing the code pointers, turns out this is a different problem than I thought.

The issue is that while your code reinitializes termdash [here](https://github.com/lefred/innotopgo/blob/9df1f34ccc073f4d27d9b4146a1a42f2a87d530d/innotop/processlist.go#L601-L604), it reuses the same instance of tcell [created here](https://github.com/lefred/innotopgo/blob/9df1f34ccc073f4d27d9b4146a1a42f2a87d530d/innotop/processlist.go#L201-L205). The tcell instance has [a goroutine](https://github.com/mum4k/termdash/blob/9ed371362bfea7c283f50f8ec783e25e7667491a/terminal/tcell/tcell.go#L128) that continues to pull keyboard events even while your program switches between the dashboards.

As a quick fix you could try closing tcell and starting a new instance of tcell every time you call termdash.Run. E.g.

```go
t, err := tcell.New()
if err != nil {
    return err
}
termdash.Run(t, ...) // Run the first dashboard.

t.Close() // Close the first tcell instance and create a new one.
t, err = tcell.New()
if err != nil {
    return err
}
termdash.Run(t, ...) // Run the second dashboard with a new instance of tcell.
```

However I would like to point out that this isn't the most straightforward way of using termdash and if you start restarting tcell as well, you might get some unwanted terminal flickering. The way termdash was intended to be used was to only call `termdash.Run` once for the lifetime of the application. And when you need to change dashboards, you can use the container.Update method. If you also need to change the behavior of the `KeyboardSubscriber`, you can define it as a method on an object in your program and that way modify its behavior without calling termdash.Run multiple times.

Please let me know if this helps.
